types/bluebird/index.d.ts

```diff
@@ -37,25 +37,32 @@
 
 type CatchFilter<E> = (new (...args: any[]) => E) | ((error: E) => boolean) | (object & E);
 type IterableItem<R> = R extends Iterable<infer U> ? U : never;
-type IterableOrNever<R> = R extends Iterable<any> ? R : never;
+type IterableOrNever<R> = Extract<R, Iterable<any>>;
+type Resolvable<R> = R | PromiseLike<R>;
+type ExtractFunction<T> = Extract<T, (...args: any[]) => any>;
+type IterateFunction<T, R> = (item: T, index: number, arrayLength: number) => Resolvable<R>;
```

Comment:
Bluebird calls this function "mapper". See [here](http://bluebirdjs.com/docs/api/promise.map.html)
Should we thus call this type `MapperFunction` for consistency? IMO that would sound better than "iterate" too.

Reply:
`IterateFunction` is also used for [`Bluebird.filter()`](http://bluebirdjs.com/docs/api/promise.filter.html) and [`Bluebird.each()`](http://bluebirdjs.com/docs/api/promise.each.html), each with a different definition name.
The most universal name is `iterator`, which conflicts with the `Iterator Protocol`.
IMO, if you wish to add an alias, you will use `Iterate`.

Reply:
Hmm fair enough.
Altough I do have to say that `IterateFunction` sounds weird in english.
Reason being: `Iterate` is a verb, while e.g. `Mapper`, `Iterator`, `Resolvable`, etc. are either objects or adjectives.
If we can find a fitting object/adjective as the prefix I'd be more in favor.

Either way you may see this particular comment as optional now.
I'm okay-ish if we keep it as `IterateFunction`.